File: python/questions/getIntersectionNode.py
from typing import Optional

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        cA = headA
        cB = headB
        while cA and cB:
            cA = cA.next
            cB = cB.next
        if cB == None:
            headA, headB = headB, headA
            cA, cB = cB, cA
        while cB:
            cB = cB.next
            headB = headB.next
        while headA != headB:
            headA = headA.next
            headB = headB.next
        return headA

File: cpp/questions/getIntersectionNode.cpp
#include <algorithm>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
        ListNode* cA = headA;
        ListNode* cB = headB;
        while (cA && cB) {
            cA = cA->next;
            cB = cB->next;
        }
        if (cB == nullptr) {
            std::swap(headA, headB);
            std::swap(cA, cB);
        }
        while (cB) {
            cB = cB->next;
            headB = headB->next;
        }
        while (headA != headB) {
            headA = headA->next;
            headB = headB->next;
        }
        return headA;
    }
};
